### PR TITLE
Support positional Constants in Wrappers

### PR DESCRIPTION
**Pull request by elad-c**
## Pull Request Description:
This PR adds the ability to quantize constants that are not attributes of layers (e.g. quantize the cosntant 2.9965 in y=tf.add(x, 2.9965), both in Keras and PyTorch.
Checkout the documentations of KerasQuantizationWrapper and PytorchQuantizationWrapper for more information.

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: mct_quantizers/pytorch/quantize_wrapper.py [hunk 1]
```diff
@@ -27,14 +28,54 @@
...
+                >>> import torch
+
+                >>> attr_quant_dict = {'weight': mctq.pytorch.quantizers.WeightsPOTInferableQuantizer(4, [2.0], False)}
+                >>> QuantizedConv2D = mctq.PytorchQuantizationWrapper(torch.nn.Conv2d(3, 3, 3), attr_quant_dict)
+
+                creating a quantized function with a constant: torch.sub(constant, Tensor)
+
+                >>> attr_quant_dict = {0: mctq.pytorch.quantizers.WeightsPOTInferableQuantizer(4, [2.0], False)}
+                >>> attr_values = {0: torch.Tensor([1, 2, 3], dtype=torch.float32)}
+                >>> QuantizedConv2D = mctq.PytorchQuantizationWrapper(torch.sub), attr_quant_dict, attr_values)
```
Comment (Chizkiyahu) [inline mct_quantizers/pytorch/quantize_wrapper.py:69]: fix name

File: mct_quantizers/pytorch/quantize_wrapper.py [hunk 2]
```diff
@@ -27,14 +28,54 @@
...
+
+                creating a quantized function with constants and arguments: tf.cat([constant#1, Tensor, constant#2], dim=1)
+                >>> attr_quant_dict = {0: mctq.pytorch.quantizers.WeightsPOTInferableQuantizer(4, [2.0], False),
+                >>>                    2: mctq.pytorch.quantizers.WeightsPOTInferableQuantizer(4, [1.0], False)}
+                >>> attr_values = {0: torch.Tensor([[1,2,3], [4, 5, 6]], dtype=torch.float32),
+                >>>                2: torch.Tensor([[4,5,6], [4, 5, 6]], dtype=torch.float32)}
+                >>> QuantizedConv2D = mctq.PytorchQuantizationWrapper(torch.cat, attr_quant_dict, attr_values,
```
Comment (Chizkiyahu) [inline mct_quantizers/pytorch/quantize_wrapper.py:76]: fix name

**Review by Chizkiyahu — CHANGES_REQUESTED** [review]

**Review by ofirgo — APPROVED** [review]
Looks great!

Please address the several minor comments that I left, but everything seems fine so I'm approving the PR.

Maybe consider updating our Readme (in a separate PR) with a section explaining the wrapper in details, and referring to the two options of regular weights and positional weights.

File: mct_quantizers/keras/quantize_wrapper.py [hunk 3]
```diff
@@ -153,17 +253,38 @@ def from_config(cls, config):
...
 
             """
+            numpy_deserialization = lambda **_config: tf.constant(**_config).numpy()
+            tensor_deserialization = lambda **_config: tf.constant(**_config)
+
+            def maybe_int(x):
+                try:
+                    return int(x)
+                except ValueError:
+                    return x
```
Comment (reuvenperetz) [inline mct_quantizers/keras/quantize_wrapper.py:263]: If it's aimed to convert strings only, maybe use isdigit instead of relying on exceptions for the program flow?

File: mct_quantizers/keras/quantize_wrapper.py [hunk 4]
```diff
@@ -110,9 +192,13 @@ def get_config(self):
 
             """
             base_config = super(KerasQuantizationWrapper, self).get_config()
-            config = {WEIGHTS_QUANTIZERS: {k: keras.utils.serialize_keras_object(v) for k, v in self.weights_quantizers.items()}}
+            config = {WEIGHTS_QUANTIZERS: {k: keras.utils.serialize_keras_object(v) for k, v in self.weights_quantizers.items()},
```
[the remark this answers is not in the document]
Comment (elad-c) [reply mct_quantizers/keras/quantize_wrapper.py:195]: yes

File: tests/compatibility_tests/keras_comp_tests/base_weights_compatibility_test.py [hunk 5]
```diff
@@ -156,8 +156,7 @@ def build_and_save_model(self, quantizer, quantizer_params, layer, model_name, i
 
         weights_quantizer = quantizer(**quantizer_params)
 
-        quant_weights_layer = KerasQuantizationWrapper(layer)
-        quant_weights_layer.add_weights_quantizer(weight_name, weights_quantizer)
+        quant_weights_layer = KerasQuantizationWrapper(layer, {weight_name: weights_quantizer})
```
[the remark this answers is not in the document]
Comment (elad-c) [reply tests/compatibility_tests/keras_comp_tests/base_weights_compatibility_test.py:159]: Seems to be the same

File: mct_quantizers/keras/quantize_wrapper.py [hunk 6]
```diff
@@ -52,40 +55,119 @@ def _weight_name(name: str) -> str:
...
+            self.is_inputs_as_list = is_inputs_as_list
+
+            # Sanity checks
+            if len(self.weight_values) == 0:
+                # expecting weights_quantizers keys to be all strings
+                if not all([isinstance(w, str) for w in self.weights_quantizers]):
+                    Logger.error('"weights_quantizers" keys should be all strings')
+                self.is_str_attr = True
+            else:
+                # expecting both weights_quantizers and weight_values keys to be all integers
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply mct_quantizers/keras/quantize_wrapper.py:160]: Maybe add a comment stating this assumption (declaring that there is no support for mixed weights layers so it would be clear)

**Review by Chizkiyahu — APPROVED** [review]

File: mct_quantizers/pytorch/quantize_wrapper.py [hunk 7]
```diff
@@ -43,42 +84,59 @@ def __init__(self,
...
+                # expecting both weights_quantizers and weight_values keys to be all integers.
+                if not all([isinstance(w, int) for w in self.weight_values]):
+                    Logger.error('All "weight_values" keys should be integers')
+                if not all([a == b for a, b in zip(weights_quantizers, weight_values)]):
+                    Logger.error('Mismatch between "weights_quantizers" and "weight_values" keys')
+                self.is_str_attr = False
 
-            """
-            self.weights_quantizers.update({param_name: quantizer})
+            self._set_weights_vars(True)
```
Comment (reuvenperetz) [inline mct_quantizers/pytorch/quantize_wrapper.py:118]: Please add a comment why is_training should be set to True since it may confuse

File: tests/pytorch_tests/test_pytorch_quantization_wrapper.py [hunk 8]
```diff
@@ -62,20 +63,67 @@ def initialize_quantization(self, tensor_shape, name, layer):
...
+        self.assertTrue([wv[0] for wv in wrapper._weights_vars] == [0, 2])
+        self.assertTrue((weight == getattr(wrapper, f'{POSITIONAL_WEIGHT}_{name}')).all())
+        self.assertTrue(isinstance(quantizer, ZeroWeightsQuantizer))
+        y = wrapper(self.inputs)  # apply the wrapper on some random inputs
+        self.assertTrue((0 == getattr(wrapper, f'{QUANTIZED_POSITIONAL_WEIGHT}_0')).all())  # check the weight are now quantized
+        self.assertTrue((0 == getattr(wrapper, f'{QUANTIZED_POSITIONAL_WEIGHT}_2')).all())  # check the weight are now quantized
+        self.assertTrue((y == self.layers[2]([torch.zeros_like(self.cat_const1),
+                                              self.inputs,
+                                              torch.zeros_like(self.cat_const2)],
+                                             **wrapper.op_call_kwargs)).all())  # check the wrapper's outputs are equal to biases
```
Comment (reuvenperetz) [inline tests/pytorch_tests/test_pytorch_quantization_wrapper.py:129]: Add a test to check if it works properly on the ONNX exported functional layer.
